### ISSUE TITLE
Nuclear weapon statistics using monoids groups and modules in Haskell

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2974]
Nuclear weapon statistics using monoids, groups, and modules in Haskell<br>
<br>
https://ift.tt/2aDCdzh<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div id="content">
<h1>Nuclear weapon statistics using monoids, groups, and modules in Haskell</h1>
<p>posted on 2013-01-04</p>
<p><img src="http://upload.wikimedia.org/wikipedia/commons/thumb/7/79/Operation_Upshot-Knothole_-_Badger_001.jpg/282px-Operation_Upshot-Knothole_-_Badger_001.jpg"></p>
<p>The <a href="http://www.thebulletin.org/">Bulletin of the Atomic Scientists</a> tracks the nuclear capabilities of every country. We&rsquo;re going to use their data to demonstrate Haskell&rsquo;s <a href="http://hackage.haskell.org/package/HLearn-algebra">HLearn</a> library and the usefulness of abstract algebra to statistics. Specifically, we&rsquo;ll see that the <a href="https://en.wikipedia.org/wiki/Categorical_distribution">categorical distribution</a> and <a href="https://en.wikipedia.org/wiki/Kernel_density_estimation">kernel density estimates</a> have <a href="https://en.wikipedia.org/wiki/Monoid">monoid</a>, <a href="https://en.wikipedia.org/wiki/Group_(mathematics)">group</a>, and <a href="https://en.wikipedia.org/wiki/Module_(mathematics)">module</a> algebraic structures. &nbsp;We&rsquo;ll explain what this crazy lingo even means, then take advantage of these structures to <strong>efficiently</strong> <strong>answer real-world statistical questions about nuclear war</strong>. It&rsquo;ll be a <a href="https://en.wikipedia.org/wiki/WOPR">WOPR</a>!</p>
<p>Before we get into the math, we&rsquo;ll need to review the basics of nuclear politics.</p>
<p>The nuclear <a href="https://en.wikipedia.org/wiki/Nuclear_Non-Proliferation_Treaty">Non-Proliferation Treaty</a> (<strong>NPT</strong>) is the main treaty governing nuclear weapons. Basically, it says that there are five countries that are &ldquo;allowed&rdquo; to have nukes: the <strong>USA</strong>, <strong>UK</strong>, <strong>France</strong>, <strong>Russia</strong>, and <strong>China</strong>. &ldquo;Allowed&rdquo; is in quotes because the treaty specifies that these countries must eventually get rid of their nuclear weapons at some future, unspecified date. When another country, for example Iran, signs the NPT, they are agreeing to not develop nuclear weapons. What they get in exchange is help from the 5 nuclear weapons states in developing their own civilian nuclear power programs. (Iran has the legitimate complaint that Western countries are actively trying to stop its civilian nuclear program when they&rsquo;re supposed to be helping it, but that&rsquo;s a <a href="http://www.csmonitor.com/Commentary/Opinion/2010/0917/Reality-check-Iran-is-not-a-nuclear-threat">whole &rsquo;nother can of worms</a>.)</p>
<p>The</p>
<a href="http://bos.sagepub.com/">Nuclear Notebook</a>
<p>tracks the nuclear capabilities of all these countries. &nbsp;The most-current estimates are from mid-2012. &nbsp;Here&rsquo;s a summary (click the warhead type for more info):</p>
<table align="center" border="1">
<tbody>
<tr>
<td align="left"><strong>Country</strong></td>
<td align="left"><strong>Delivery Method</strong></td>
<td align="left"><strong>Warhead</strong></td>
<td align="left"><strong>Yield (kt)</strong></td>
<td align="left"><strong># Deployed</strong></td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">ICBM</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W78.html">W78</a></td>
<td align="right">335</td>
<td align="right">250</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">ICBM</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W87.html">W87</a></td>
<td align="right">300</td>
<td align="right">250</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">SLBM</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W76.html">W76</a></td>
<td align="right">100</td>
<td align="right">468</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">SLBM</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W76.html">W76-1</a></td>
<td align="right">100</td>
<td align="right">300</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">SLBM</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W88.html">W88</a></td>
<td align="right">455</td>
<td align="right">384</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">Bomber</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W80.html">W80</a></td>
<td align="right">150</td>
<td align="right">200</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">Bomber</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/B61_nuclear_bomb">B61</a></td>
<td align="right">340</td>
<td align="right">50</td>
</tr>
<tr>
<td align="left">USA</td>
<td align="left">Bomber</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/B83_nuclear_bomb">B83</a></td>
<td align="right">1200</td>
<td align="right">50</td>
</tr>
<tr>
<td align="left">UK</td>
<td align="left">SLBM</td>
<td align="left"><a href="http://nuclearweaponarchive.org/Usa/Weapons/W76.html">W76</a></td>
<td align="right">100</td>
<td align="right">225</td>
</tr>
<tr>
<td align="left">France</td>
<td align="left">SLBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/TN_75">TN75</a></td>
<td align="right">100</td>
<td align="right">150</td>
</tr>
<tr>
<td align="left">France</td>
<td align="left">Bomber</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/TN_81">TN81</a></td>
<td align="right">300</td>
<td align="right">150</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">ICBM</td>
<td align="left">RS-20V</td>
<td align="right">800</td>
<td align="right">500</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">ICBM</td>
<td align="left">RS-18</td>
<td align="right">400</td>
<td align="right">288</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">ICBM</td>
<td align="left">RS-12M</td>
<td align="right">800</td>
<td align="right">135</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">ICBM</td>
<td align="left">RS-12M2</td>
<td align="right">800</td>
<td align="right">56</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">ICBM</td>
<td align="left">RS-12M1</td>
<td align="right">800</td>
<td align="right">18</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">ICBM</td>
<td align="left">RS-24</td>
<td align="right">100</td>
<td align="right">90</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">SLBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/R-29_Vysota">RSM-50</a></td>
<td align="right">50</td>
<td align="right">144</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">SLBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/R-29RM_Shtil">RSM-54</a></td>
<td align="right">100</td>
<td align="right">384</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="left">Bomber</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/Kh-55_(missile_family)">AS-15</a></td>
<td align="right">200</td>
<td align="right">820</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">ICBM</td>
<td align="left"><a href="https://www.fas.org/nuke/guide/china/theater/df-3a.htm">DF-3A</a></td>
<td align="right">3300</td>
<td align="right">16</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">ICBM</td>
<td align="left"><a href="https://www.fas.org/nuke/guide/china/theater/df-4.htm">DF-4</a></td>
<td align="right">3300</td>
<td align="right">12</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">ICBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/DF-5">DF-5A</a></td>
<td align="right">5000</td>
<td align="right">20</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">ICBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/DF-21">DF-21</a></td>
<td align="right">300</td>
<td align="right">60</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">ICBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/DF-31">DF-31</a></td>
<td align="right">300</td>
<td align="right">20</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">ICBM</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/DF-31">DF-31A</a></td>
<td align="right">300</td>
<td align="right">20</td>
</tr>
<tr>
<td align="left">China</td>
<td align="left">Bomber</td>
<td align="left"><a href="https://en.wikipedia.org/wiki/Xian_H-6">H-6</a></td>
<td align="right">3100</td>
<td align="right">20</td>
</tr>
</tbody>
</table>
<p>I&rsquo;ve consolidated all this data into the file <a href="http://izbicki.me/public/datasets/nukes-list.csv">nukes-list.csv</a>, which we will analyze in this post. &nbsp;If you want to try out this code for yourself (or the homework question at the end), you&rsquo;ll need to download it. &nbsp;Every line in the file corresponds to a single nuclear warhead, not delivery method. &nbsp;Warheads are the parts that go boom! &nbsp;Bombers, <a href="https://en.wikipedia.org/wiki/Icbm">ICBMs</a>, and <a href="https://en.wikipedia.org/wiki/SSBN">SSBN</a>/<a href="https://en.wikipedia.org/wiki/SLBMs">SLBMs</a> are the delivery method.</p>
<div>
<img src="https://izbicki.me/img/uploads/2013/01/nuclear-triad.jpg" alt="nuclear-triad">
<p>nuclear-triad</p>
</div>
<p>There are three things to note about this data. &nbsp;First, it&rsquo;s <strong>only estimates</strong> based on public sources. &nbsp;In particular, it probably overestimates the Russian nuclear forces. <a href="http://russianforces.org/missiles/">Other estimates are considerably lower</a>. &nbsp;Second, we will only be considering <strong>deployed, strategic warheads</strong>. &nbsp;Basically, this means the &ldquo;really big nukes that are currently aimed at another country.&rdquo; &nbsp;There are thousands more tactical warheads, and warheads in reserve stockpiles waiting to be disassembled. &nbsp;For simplicity&mdash;and because these nukes don&rsquo;t significantly affect strategic planning&mdash;we won&rsquo;t be considering them here. &nbsp; Finally, there are 4 countries who are not members of the NPT but have nuclear weapons:&nbsp;<strong>Israel</strong>, <strong>Pakistan</strong>, <strong>India</strong>, and <strong>North Korea</strong>. &nbsp;We will be ignoring them here because their inventories are relatively small, and most of their weapons would not be considered strategic.</p>
<h3 id="programming-preliminaries">Programming preliminaries</h3>
<p>First, let&rsquo;s install the library:</p>
<pre>
<code>$ cabal install HLearn-distributions-0.1</code>
</pre>
<p>Now we&rsquo;re ready to start programming. First, let&rsquo;s import our libraries:</p>
<pre>
<code>&gt;import Control.Lens &gt;import Data.Csv &gt;import qualified Data.Vector as V &gt;import qualified Data.ByteString.Lazy.Char8 as BS &gt; &gt;import HLearn.Algebra &gt;import HLearn.Models.Distributions &gt;import HLearn.Gnuplot.Distributions</code>
</pre>
<p>Next, we load our data using the <a href="http://hackage.haskell.org/package/cassava">Cassava</a> package. &nbsp;(You don&rsquo;t need to understand how this works.)</p>
<pre>
<code>&gt;main = do &gt; Right rawdata &lt;- fmap (fmap V.toList . decode True) $ BS.readFile "nukes-list.csv" &gt;        :: IO (Either String [(String, String, String, Int)])</code>
</pre>
<p>And we&rsquo;ll use the <a href="http://hackage.haskell.org/package/lens">Lens</a> package to parse the CSV file into a series of variables containing just the values we want. &nbsp;(You also don&rsquo;t need to understand this.)</p>
<pre>
<code>&gt; let list_usa = fmap (\row -&gt; row^._4) $ filter (\row -&gt; (row^._1)=="USA" ) rawdata &gt; let list_uk = fmap (\row -&gt; row^._4) $ filter (\row -&gt; (row^._1)=="UK" ) rawdata &gt; let list_france = fmap (\row -&gt; row^._4) $ filter (\row -&gt; (row^._1)=="France") rawdata &gt; let list_russia = fmap (\row -&gt; row^._4) $ filter (\row -&gt; (row^._1)=="Russia") rawdata &gt; let list_china = fmap (\row -&gt; row^._4) $ filter (\row -&gt; (row^._1)=="China" ) rawdata</code>
</pre>
<p><strong>NOTE:</strong> All you need to understand about the above code is what these list_country variables look like. So let&rsquo;s print one:</p>
<pre>
<code>&gt; putStrLn $ "List of American nuclear weapon sizes = " ++ show list_usa</code>
</pre>
<p>gives us the output:</p>
<pre>
<code>List of American nuclear weapon sizes = fromList [335,335,335,335,335,335,335,335,335,335 ... 1200,1200,1200,1200,1200]</code>
</pre>
<p>If we want to know how many weapons are in the American arsenal, we can take the length of the list:</p>
<pre>
<code>&gt; putStrLn $ "Number of American weapons = " ++ show (length list_usa)</code>
</pre>
<p>We get that there are <strong>1951 American deployed, strategic nuclear weapons</strong>. &nbsp;If we want to know the total &ldquo;blowing up&rdquo; power, we take the sum of the list:</p>
<pre>
<code>&gt; putStrLn $ "Explosive power of American weapons = " ++ show (sum list_usa)</code>
</pre>
<p>We get that the US has &nbsp;<strong>516 megatons of deployed, strategic nuclear weapons</strong>. &nbsp;That&rsquo;s the equivalent of&nbsp;<strong>1,033,870,000,000 pounds of TNT</strong>.</p>
<p>To get the total number of weapons in the world, we concatenate every country&rsquo;s list of weapons and find the length:</p>
<pre>
<code>&gt; let list_all = list_usa ++ list_uk ++ list_france ++ list_russia ++ list_china &gt; putStrLn $ "Number of nukes in the whole world = " ++ show (length list_all)</code>
</pre>
<p>Doing this for every country gives us the table:</p>
<table align="center" border="1">
<tbody>
<tr>
<td align="left"><strong>Country</strong></td>
<td align="left"><strong>Warheads</strong></td>
<td align="left"><strong>Total explosive power (kt)</strong></td>
</tr>
<tr>
<td align="left">USA</td>
<td align="right">1,951</td>
<td align="right">516,935</td>
</tr>
<tr>
<td align="left">UK</td>
<td align="right">225</td>
<td align="right">22,500</td>
</tr>
<tr>
<td align="left">France</td>
<td align="right">300</td>
<td align="right">60,000</td>
</tr>
<tr>
<td align="left">Russia</td>
<td align="right">2,435</td>
<td align="right">901,000</td>
</tr>
<tr>
<td align="left">China</td>
<td align="right">168</td>
<td align="right">284,400</td>
</tr>
<tr>
<td align="left"><strong>Total</strong></td>
<td align="right">5,079</td>
<td align="right">1,784,835</td>
</tr>
</tbody>
</table>
<p>Now let&rsquo;s do some algebra!</p>
<h3 id="monoids-and-groups">Monoids and groups</h3>
<p>In a previous post, we saw that the <a href="http://izbicki.me/blog/gausian-distributions-are-monoids">Gaussian distribution forms a group</a>. This means that it has all the properties of a monoid&mdash;an empty element (<strong>mempty</strong>) that represents the distribution trained on no data, and a binary operation (<strong>mappend</strong>) that merges two distributions together&mdash;plus an <strong>inverse</strong>. This inverse lets us &ldquo;subtract&rdquo; two Gaussians from each other.</p>
<p>It turns out that many other distributions also have this group property. For example, the <strong>categorical distribution.</strong>&nbsp;&nbsp;This distribution is used for measuring discrete data. Essentially, it assigns some probability to each &ldquo;label.&rdquo; &nbsp;In our case, the labels are the size of the nuclear weapon, and the probability is the chance that a randomly chosen nuke will be exactly that destructive. &nbsp;We train our categorical distribution using the <a href="http://hackage.haskell.org/packages/archive/HLearn-algebra/0.0.1/doc/html/HLearn-Algebra-Models.html">train</a> function:</p>
<pre>
<code>&gt; let cat_usa = train list_usa :: Categorical Int Double</code>
</pre>
<p>If we plot this distribution, we&rsquo;ll get a graph that looks something like:</p>
<center><img src="https://izbicki.me/img/uploads/2013/01/catigorical-distribution-of-american-nuclear-weapons1.png" alt="catigorical distribution of american nuclear weapons"></center>
<p>A distribution like this is useful to war planners from other countries. &nbsp;It can help them statistically determine the amount of casualties their infrastructure will take from a nuclear exchange.</p>
<p>Now, let&rsquo;s train equivalent distributions for our other countries.</p>
<pre>
<code>&gt; let cat_uk = train list_uk :: Categorical Int Double &gt; let cat_france = train list_france :: Categorical Int Double &gt; let cat_russia = train list_russia :: Categorical Int Double &gt; let cat_china = train list_china :: Categorical Int Double</code>
</pre>
<p>Because training the categorical distribution is a group <strong>homomorphism</strong>, we can train a distribution over all nukes by either training directly on the data:</p>
<pre>
<code>&gt; let cat_allA = train list_all :: Categorical Int Double</code>
</pre>
<p>or we can merge the already generated categorical distributions:</p>
<pre>
<code>&gt; let cat_allB = cat_usa &lt;&gt; cat_uk &lt;&gt; cat_france &lt;&gt; cat_russia &lt;&gt; cat_china</code>
</pre>
<p>Because of the homomorphism property, we will get the same result both ways. Since we&rsquo;ve already done the calculations for each of the the countries already, method B will be more efficient&mdash;it won&rsquo;t have to repeat work we&rsquo;ve already done. &nbsp;If we plot either of these distributions, we get:</p>
<center><img src="https://izbicki.me/img/uploads/2013/01/catigorical-distribution-of-all-nuclear-weapons.png" alt="catigorical distribution of all nuclear weapons"></center>
<p>The thing to notice in this plot is that most countries have a nuclear arsenal that is distributed similarly to the United States&mdash;except for China. &nbsp;These Chinese ICBMs will become much more important when we discuss nuclear strategy in the last section.</p>
<p>But nuclear war planners don&rsquo;t particularly care about this complete list of nuclear weapons. &nbsp;What war planners care about is the <strong>survivable nuclear weapons</strong>&mdash;that is, weapons that won&rsquo;t be blown up by a surprise nuclear attack. &nbsp;Our distributions above contain nukes dropped from bombers, but these are not survivable. &nbsp;They are easy to destroy. &nbsp;For our purposes, we&rsquo;ll call anything that&rsquo;s not a bomber a survivable weapon.</p>
<div>
<img src="https://izbicki.me/img/uploads/2013/01/nuclear-triad-no-bomber.jpg" alt="nuclear-triad-no-bomber">
<p>nuclear-triad-no-bomber</p>
</div>
<p>We&rsquo;ll use the group property of the categorical distribution to calculate the survivable weapons. &nbsp;First, we create a distribution of just the _un_survivable bombers:</p>
<pre>
<code>&gt; let list_bomber = fmap (\row -&gt; row^._4) $ filter (\row -&gt; (row^._2)=="Bomber") rawdata &gt; let cat_bomber = train list_bomber :: Categorical Int Double</code>
</pre>
<p>Then, we use our group inverse to subtract these unsurvivable weapons away:</p>
<pre>
<code>&gt; let cat_survivable = cat_allB &lt;&gt; (inverse cat_bomber)</code>
</pre>
<p>Notice that we calculated this distribution indirectly&mdash;there was no possible way to combine our variables above to generate this value without using the inverse! This is the power of groups in statistics.</p>
<h3 id="more-distributions">More distributions</h3>
<p>The categorical distribution is not sufficient to accurately describe the distribution of nuclear weapons. This is because we don&rsquo;t actually know the yield of a given warhead. Like all things, it has some manufacturing tolerances that we must consider. For example, if we detonate a 300 kt warhead, the actual explosion might be 275 kt, 350 kt, or the bomb might even &ldquo;fizzle out&rdquo; and have almost a 0kt explosion.</p>
<p>We&rsquo;ll model this by using a <strong>kernel density estimator</strong> (KDE). &nbsp;The KDE basically takes all our data points, assigns each one a probability distribution called a &ldquo;kernel,&rdquo; then sums these kernels together. &nbsp;It is a very powerful and general technique for modelling distributions&hellip; and it also happens to form a group!</p>
<p>First, let&rsquo;s create the parameters for our KDE. &nbsp;The bandwidth controls how wide each of the kernels is. &nbsp;Bigger means wider. &nbsp;I selected 20 because it made a reasonable looking density function. &nbsp;The sample points are exactly what they sounds like: they are where we will sample the density from. &nbsp;We can generate them using the function <a href="http://hackage.haskell.org/packages/archive/HLearn-distributions/0.1.0.1/doc/html/HLearn-Models-Distributions-KernelDensityEstimator.html#g:3">genSamplePoints</a>. &nbsp;Finally, the kernel is the shape of the distributions we will be summing up. &nbsp;There are many <a href="http://hackage.haskell.org/packages/archive/HLearn-distributions/0.1.0.1/doc/html/HLearn-Models-Distributions-KernelDensityEstimator-Kernels.html">supported kernels</a>.</p>
<pre>
<code>&gt; let kdeparams = KDEParams &gt; { bandwidth = Constant 20 &gt; , samplePoints = genSamplePoints &gt; 0 -- minimum &gt; 4000 -- maximum &gt; 4000 -- number of samples &gt; , kernel = KernelBox Gaussian &gt; } :: KDEParams Double</code>
</pre>
<p>Now, we&rsquo;ll train kernel density estimates on our data. &nbsp;Notice that because the KDE takes parameters, we must use the <strong>train&rsquo;</strong> function instead of just train.</p>
<pre>
<code>&gt; let kde_usa = train' kdeparams list_usa :: KDE Double</code>
</pre>
<p>Again, plotting just the American weapons gives:</p>
<div>
<img src="https://izbicki.me/img/uploads/2013/01/kernel-density-estimate-of-american-nuclear-weapons.png" alt="kernel density estimate of american nuclear weapons">
<p>kernel density estimate of american nuclear weapons</p>
</div>
<p>And we train the corresponding distributions for the other countries.</p>
<pre>
<code>&gt; let kde_uk = train' kdeparams list_uk :: KDE Double &gt; let kde_france = train' kdeparams list_france :: KDE Double &gt; let kde_russia = train' kdeparams list_russia :: KDE Double &gt; let kde_china = train' kdeparams list_china :: KDE Double &gt; &gt; let kde_all = kde_usa &lt;&gt; kde_uk &lt;&gt; kde_france &lt;&gt; kde_russia &lt;&gt; kde_china</code>
</pre>
<p>The KDE is a powerful technique, but the draw back is that it is computationally expensive&mdash;especially when a large number of sample points are used. Fortunately, all computations in the HLearn library are <strong>easily parallelizable</strong> by applying the higher order function <a href="http://hackage.haskell.org/packages/archive/HLearn-algebra/0.0.1/doc/html/HLearn-Algebra-Functions.html">parallel</a>.</p>
<p>We can calculate the full KDE from scratch in parallel like this:</p>
<pre>
<code>&gt; let list_double_all = map fromIntegral list_all :: [Double] &gt; let kde_all_parA = (parallel (train' kdeparams)) list_double_all :: KDE Double</code>
</pre>
<p>or we can perform a parallel reduction on the KDEs for each country like this:</p>
<pre>
<code>&gt; let kde_all_parB = (parallel reduce) [kde_usa, kde_uk, kde_france, kde_russia, kde_china]</code>
</pre>
<p>And because the KDE is a homomorphism, we get the same exact thing either way. &nbsp;Let&rsquo;s plot the parallel version:</p>
<pre>
<code>&gt; plotDistribution (genPlotParams "kde_all" kde_all_parA) kde_all_parA</code>
</pre>
<center><img src="https://izbicki.me/img/uploads/2013/01/kernel-density-estimate-of-all-nuclear-weapons-globally-mod.png" alt="kernel density estimate of all nuclear weapons globally-mod"></center>
<p>The parallel computation takes about 16 seconds on my Core2 Duo laptop running on 2 processors, whereas the serial computation takes about 28 seconds.</p>
<p>This is a considerable speedup, but we can still do better. It turns out that there is a homomorphism from the Categorical distribution to the KDE:</p>
<pre>
<code>&gt; let kde_fromcat_all = cat_allB $&gt; kdeparams &gt; plotDistribution (genPlotParams "kde_fromcat_all" kde_fromcat_all) kde_fromcat_all</code>
</pre>
<p>(For more information about the morphism chaining operator&nbsp;<strong>$&gt;</strong>, see the <a href="http://hackage.haskell.org/packages/archive/HLearn-algebra/0.1.0.1/doc/html/HLearn-Algebra-Morphism.html">Hlearn documentation</a>.)&nbsp;This computation takes less than a second and gets the exact same result as the much more expensive computations above.</p>
<p>We can express this relationship with a commutative diagram:</p>
<center><img src="https://izbicki.me/img/uploads/2013/01/kde-commutative-diagram-small.png" alt="kde-commutative-diagram-small"></center>
<p>No matter which path we take to get to a KDE, we will get the exact same answer. &nbsp;So we should always take the path that will be least computationally expensive for the data set we&rsquo;re working on.</p>
<p>Why does this work? Well, the categorical distribution is a structure called the &ldquo;free module&rdquo; in disguise.</p>
<h3 id="modules-and-the-free-module">Modules and the Free Module</h3>
<p><strong>R-Modules</strong> (like groups, but unlike monoids) have not seen much love from functional programmers. This is a shame, because they&rsquo;re quite handy. It turns out they will increase our performance dramatically in this case.</p>
<p>It&rsquo;s not super important to know the formal definition of an R-module, but here it is anyways: An R-module is a group with an additional property: it can be &ldquo;multiplied&rdquo; by any element of the <a href="https://en.wikipedia.org/wiki/Ring_(mathematics)">ring</a> R. This is a generalization of <a href="https://en.wikipedia.org/wiki/Vector_space">vector spaces</a> because R need only be a ring instead of a <a href="https://en.wikipedia.org/wiki/Field_(mathematics)">field</a>. (Rings do not necessarily have multiplicative inverses.) &nbsp;It&rsquo;s probably easier to see what this means by an example.</p>
<p>Vectors are modules. &nbsp;Let&rsquo;s say I have a vector:</p>
<pre>
<code>&gt; let vec = [1,2,3,4,5] :: [Int]</code>
</pre>
<p>I can perform scalar multiplication on that vector like this:</p>
<pre>
<code>&gt; let vec2 = 3 .* vec</code>
</pre>
<p>which as you might expect results in:</p>
<pre>
<code>[3,6,9,12,15]</code>
</pre>
<p>Our next example is the <strong>free R-module</strong>. A &ldquo;free&rdquo; structure is one that obeys only the axioms of the structure and nothing else. Functional programmers are very familiar with the free monoid&mdash;it&rsquo;s the list data type. The <strong>free Z-module</strong> is like a beefed up list. Instead of just storing the elements in a list, it also stores the number of times that element occurred. &nbsp;(Z is shorthand for the set of integers, which form a ring but not a field.) This lets us greatly reduce the memory required to store a repetitive data set.</p>
<p>In HLearn, we represent the free module over a ring r with the data type:</p>
<pre>
<code>:: FreeMod r a</code>
</pre>
<p>where a is the type of elements to be stored in the free module. We can convert our lists into free modules using the function <strong>list2module</strong> like this:</p>
<pre>
<code>&gt; let module_usa = list2module list_usa</code>
</pre>
<p>But what does the free module actually look like? Let&rsquo;s print it to find out:</p>
<pre>
<code>&gt; print module_usa</code>
</pre>
<p>gives us:</p>
<pre>
<code>FreeMod (fromList [(100,768),(150,200),(300,250),(335,249),(340,50),(455,384),(1200,50)])</code>
</pre>
<p>This is much more compact! So this is the take away: <strong>The free module makes repetitive data sets easier to work with.</strong> Now, let&rsquo;s convert all our country data into module form:</p>
<pre>
<code>&gt; let module_uk = list2module list_uk &gt; let module_france = list2module list_france &gt; let module_russia = list2module list_russia &gt; let module_china = list2module list_china</code>
</pre>
<p>Because modules are also groups, we can combine them like so:</p>
<pre>
<code>&gt; let module_allA = module_usa &lt;&gt; module_uk &lt;&gt; module_france &lt;&gt; module_russia &lt;&gt; module_china</code>
</pre>
<p>or, we could train them from scratch:</p>
<pre>
<code>&gt; let module_allB = list2module list_all</code>
</pre>
<p>Again, because generating a free module is a homomorphism, both methods are equivalent.</p>
<h3 id="module-distributions">Module distributions</h3>
<p>The categorical distribution and the KDE both have this module structure. This gives us two cool properties for free.</p>
<p>First, <strong>we can train these distributions directly from the free module</strong>. &nbsp;Because the free module is potentially much more compact than a list is, this can save both memory and time. If we run:</p>
<pre>
<code>&gt; let cat_module_all = train module_allB :: Categorical Int Double &gt; let kde_module_all = train' kdeparams module_allB :: KDE Double</code>
</pre>
<p>Then we get the properties:</p>
<pre>
<code>cat_mod_all == cat_all kde_mod_all == kde_all == kde_cat_all</code>
</pre>
<p>Extending our commutative diagram above gives:</p>
<center><img src="https://izbicki.me/img/uploads/2013/01/kde-commutative-diagram-big.png" alt="kde-commutative-diagram-big"></center>
<p>Again, no matter which path we take to train our KDE, we still get the same result because each of these arrows is a homomorphism.</p>
<p>Second, <strong>if a distribution is a module, we can weight the importance of our data points</strong>. &nbsp;Let&rsquo;s say we&rsquo;re a general from North Korea (DPRK), and we&rsquo;re planning our nuclear strategy. The US and North Korea have a very strained relationship in the nuclear department. It is much more likely that the US will try to nuke the DPRK than China will. And modules let us model this! &nbsp;We can weight each country&rsquo;s influence on our &ldquo;nuclear threat profile&rdquo; distribution like this:</p>
<pre>
<code>&gt; let threats_dprk = 20 .* kde_usa &gt; &lt;&gt; 10 .* kde_uk &gt; &lt;&gt; 5 .* kde_france &gt; &lt;&gt; 2 .* kde_russia &gt; &lt;&gt; 1 .* kde_china &gt; &gt; plotDistribution (genPlotParams "threats_dprk" threats_dprk) threats_dprk</code>
</pre>
<p>Basically, we&rsquo;re saying that the USA is 20x more likely to attack the DPRK than China is. &nbsp;Graphically, our threat distribution is:</p>
<center><img src="https://izbicki.me/img/uploads/2013/01/nuclear-threat-against-dprk1.png" alt="nuclear-threat-against-dprk"></center>
<p>The maximum threat that we have to worry about is about 1300 kt, so we need to <a href="https://encrypted.google.com/url?sa=t&amp;rct=j&amp;q=nuclear%20blast%20dynamics&amp;source=web&amp;cd=3&amp;ved=0CEQQFjAC&amp;url=http%3A%2F%2Fwww.dtic.mil%2Fdtic%2Ftr%2Ffulltext%2Fu2%2F601139.pdf&amp;ei=sFbmUIPtM8K6igLqpIGICg&amp;usg=AFQjCNE_yhq31Q03YvFfGl1Te_WsQlKOgw&amp;sig2=Q3YNU84DaVTtWffkOlLOCQ&amp;bvm=bv.1355534169,d.cGE">design all our nuclear bunkers to withstand this level of blast</a>. &nbsp;Nuclear war planners would use the above distribution to figure out how much infrastructure would survive a nuclear exchange. &nbsp;To see how this is done, you&rsquo;ll have to click the link.</p>
<p>On the other hand, if we&rsquo;re an <strong>American general</strong>, then we might say that China is our biggest threat&hellip; who knows what they&rsquo;ll do when we can&rsquo;t pay all the debt we owe them!?</p>
<pre>
<code>&gt; let threats_usa = 1 .* kde_russia &gt; &lt;&gt; 5 .* kde_china &gt; &gt; plotDistribution (genPlotParams "threats_usa" threats_usa) threats_usa</code>
</pre>
<p>Graphically:</p>
<center><img src="https://izbicki.me/img/uploads/2013/01/nuclear-threat-against-usa.png" alt="nuclear-threat-against-usa"></center>
<p>So now Chinese ICBMs are a real threat. &nbsp;For American infrastructure to be secure, most of it needs to be able to withstand ~3500 kt blast. &nbsp;(Actually, Chinese nuclear policy is called the &ldquo;minimum means of reprisal&rdquo;&mdash;these nukes are not targeted at military installations, but major cities. &nbsp;Unlike the other nuclear powers, China doesn&rsquo;t hope to win a nuclear war. &nbsp;Instead, its nuclear posture is designed to prevent nuclear war in the first place. &nbsp;This is why China has the fewest weapons of any of these countries. &nbsp;For a detailed analysis, see the book <a href="http://mitpress.mit.edu/books/minimum-means-reprisal">Minimum Means of Reprisal</a>. &nbsp;This means that American military infrastructure isn&rsquo;t threatened by these large Chinese nukes, and really only needs to be able to withstand an 800kt explosion to be survivable.)</p>
<p>By the way, since we&rsquo;ve already calculated all of the kde_country variables before, <strong>these computations take virtually no time at all to compute</strong>. &nbsp;Again, this is all made possible thanks to our friend abstract algebra.</p>
<h3 id="homework-next-post">Homework + next Post</h3>
<p>If you want to try out the <a href="http://hackage.haskell.org/package/HLearn-algebra">HLearn library</a> for yourself, here&rsquo;s a question you can try to answer: Create the DPRK and US threat distributions above, but only use survivable weapons. &nbsp;Don&rsquo;t include bombers in the analysis.</p>
<p>In our <strong>next post</strong>, we&rsquo;ll go into more detail about the <strong>mathematical plumbing</strong> that makes all this possible. Then we&rsquo;ll start talking about Bayesian classification and full-on machine learning. <a href="http://izbicki.me/blog/feed">Subscribe to the RSS feed</a> so you don&rsquo;t miss out!</p>
<p>Why don&rsquo;t you listen to&nbsp;<a href="http://www.youtube.com/watch?v=YDFqoReof6A">Tom Lehrer&rsquo;s &ldquo;Song for WWIII&rdquo;</a>&nbsp;while you wait?</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via izbicki.me https://izbicki.me<br>
<br>
March 12, 2019 at 03:19PM
